### PR TITLE
[MAINTENANCE] Refactor validations in Checkpoint to use `CheckpointValidationConfig`

### PR DESCRIPTION
**Pull request by cdkini**
Instead of black-box dicts, we should use the rich `CheckpointValidationConfig` type. Our APIs should support both but we should convert to the rich type under the hood so we have better type safety throughout checkpoint flows.

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more details, see our [Contribution Checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist), [Coding style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style), and [Documentation style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/docs_style).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/checkpoint/checkpoint.py [hunk 1]
```diff
@@ -773,12 +781,16 @@ def __init__(  # noqa: PLR0913
         action_list: Optional[Sequence[ActionDict]] = None,
         evaluation_parameters: Optional[dict] = None,
         runtime_configuration: Optional[dict] = None,
-        validations: Optional[List[dict]] = None,
+        validations: Optional[List[dict] | List[CheckpointValidationConfig]] = None,
```
Comment (TrangPham) [inline great_expectations/checkpoint/checkpoint.py]: nit (non-blocking): IMO This formate is more readable:
```suggestion
        validations: List[dict] | List[CheckpointValidationConfig] | None = None,
```

File: great_expectations/checkpoint/checkpoint.py [hunk 2]
```diff
@@ -853,7 +865,9 @@ def run_with_runtime_args(  # noqa: PLR0913
         action_list: Optional[Sequence[ActionDict]] = None,
         evaluation_parameters: Optional[dict] = None,
         runtime_configuration: Optional[dict] = None,
-        validations: Optional[List[dict]] = None,
+        validations: Optional[
+            Union[List[CheckpointValidationConfig], List[dict]]
+        ] = None,
```
Comment (TrangPham) [inline great_expectations/checkpoint/checkpoint.py]:
```suggestion
        validations: List[dict] | List[CheckpointValidationConfig] | None = None,
```

File: great_expectations/checkpoint/checkpoint.py [hunk 3]
```diff
@@ -1154,7 +1158,9 @@ def __init__(  # noqa: PLR0913
...
```
Comment (TrangPham) [inline great_expectations/checkpoint/checkpoint.py]:
```suggestion
        validations: List[dict] | List[CheckpointValidationConfig] | None = None,
```

File: great_expectations/checkpoint/checkpoint.py [hunk 4]
```diff
@@ -365,6 +359,20 @@ def run(  # noqa: C901, PLR0913, PLR0915
...
+    ) -> list[CheckpointValidationConfig]:
+        # We accept both dicts and rich config types but all internal usage should use the latter
+        if not validations:
+            return []
+        return [
+            CheckpointValidationConfig(**validation)
+            if isinstance(validation, dict)
+            else validation
+            for validation in validations
+        ]
```
Comment (TrangPham) [inline great_expectations/checkpoint/checkpoint.py]: (non-blocking): There are 3 instances of this block of code in this PR. Is there a way to make this more DRY?

Comment (cdkini) [reply great_expectations/checkpoint/checkpoint.py]: Yeah trying to figure out a better way but as long as this is incrementally better, I'd push to make these changes and then continue iterating with subsequent PRs - finding it quite difficult to refactor given how deeply entrenched some of these typing issues are

File: great_expectations/checkpoint/checkpoint.py (hunk 1 above)
Comment (cdkini) [reply great_expectations/checkpoint/checkpoint.py]: I agree! I think `ruff` provides an auto-fix for a lot of these so I'm just making sure the types are correct (and going back later to clean up the actual formatting). I'll make these changes before merging

File: great_expectations/checkpoint/checkpoint.py (hunk 4 above)
Comment (cdkini) [reply great_expectations/checkpoint/checkpoint.py]: moved to a util function - don't love it but it allows things to be shared. This is all a stopgap measure until we create a better Checkpoint (in `experimental`)

File: tests/checkpoint/test_checkpoint_config.py [hunk 5]
```diff
@@ -62,50 +62,19 @@ def checkpoint(
...
-    {
-      "name": "store_validation_result",
-      "action": {
-        "class_name": "StoreValidationResultAction"
-      }
-    },
-    {
-      "name": "store_evaluation_params",
-      "action": {
-        "class_name": "StoreEvaluationParametersAction"
```
Comment (cdkini) [inline tests/checkpoint/test_checkpoint_config.py:78]: The repr has differing outputs due to the variability of the `CheckpointValidationConfig`'s ordering - I figured it was easier to just test some top level keys but lmk if you disagree.

File: great_expectations/checkpoint/util.py [hunk 6]
```diff
@@ -450,39 +450,48 @@ def get_updated_action_list(
 
 
 def does_batch_request_in_validations_contain_batch_data(
-    validations: Optional[List[dict]] = None,
+    validations: List[CheckpointValidationConfig],
```
Comment (Kilo59) [inline great_expectations/checkpoint/util.py:453]: I'm guessing `CheckpointValidationConfig` is just a `DictDot`? Or does it actually declare fields and types?

File: great_expectations/checkpoint/checkpoint.py [hunk 7]
```diff
@@ -636,14 +630,14 @@ def config(self) -> CheckpointConfig:
         return self._checkpoint_config
 
     @property
-    def name(self) -> Optional[str]:
+    def name(self) -> str | None:
         try:
             return self.config.name
         except AttributeError:
             return None
```
Comment (Kilo59) [inline great_expectations/checkpoint/checkpoint.py]: Is `name` really optional/None-able?

File: great_expectations/checkpoint/checkpoint.py [hunk 8]
```diff
@@ -1143,26 +1125,24 @@ def __init__(  # noqa: PLR0913
...
         # the following four arguments are used by SimpleCheckpointConfigurator
-        site_names: Union[str, List[str]] = "all",
-        slack_webhook: Optional[str] = None,
+        site_names: str | list[str] = "all",
+        slack_webhook: str | None = None,
         notify_on: str = "all",
-        notify_with: Union[str, List[str]] = "all",
-        expectation_suite_ge_cloud_id: Optional[str] = None,
+        notify_with: str | list[str] = "all",
+        expectation_suite_ge_cloud_id: str | None = None,
```
Comment (Kilo59) [inline great_expectations/checkpoint/checkpoint.py:1141]: `non-blocking`
How many overloads would it take to cover this `__init__` method?
If it's small could you define them in this PR?

File: great_expectations/core/configuration.py [hunk 9]
```diff
@@ -16,6 +17,9 @@ def __init__(self, id: Optional[str] = None, name: Optional[str] = None) -> None
         self.name = name
         super().__init__()
 
+    def __repr__(self) -> str:
+        return json.dumps(self.to_dict(), indent=2)
```
Comment (Kilo59) [inline great_expectations/core/configuration.py]: Are we just using `json.dumps()` for the pretty formatting?
If so I would recommend using `pprint.pformat()` instead.
https://docs.python.org/3/library/pprint.html#pprint.pformat
```python
from pprint import pformat as pf

...

    def __repr__(self) -> str:
        return pf(self.to_dict(), indent=2, sort_dicts=True)
```
Can use `sort_dicts` if desired.

Also, you can take advantage of nice formatting features like `depth` which will cause the formatter to only show structures up to `x` depth, and uses `...` for everything else.
```python
>>> pf({"a": {"1": 1}, "b": {"2": 2}, "c": [1, 2, 3]}, depth=1)
... {'a': {...}, 'b': {...}, 'c': [...]}
```

**Review by Kilo59 — APPROVED** [review]
Couple non-blocking comments

File: great_expectations/checkpoint/util.py (hunk 6 above)
Comment (cdkini) [reply great_expectations/checkpoint/util.py:453]: yup its an `AbstractConfig` (which inherits from `SerializableDictDot`)

File: great_expectations/checkpoint/checkpoint.py (hunk 7 above)
Comment (cdkini) [reply great_expectations/checkpoint/checkpoint.py]: hmm no I don't think so - updating now

File: great_expectations/checkpoint/checkpoint.py (hunk 8 above)
Comment (cdkini) [reply great_expectations/checkpoint/checkpoint.py:1141]: ah I'm afraid it is quite cumbersome - I think it isn't worth the effort (but we can apply this to our updated checkpoint in experimental)

File: great_expectations/core/configuration.py (hunk 9 above)
Comment (cdkini) [reply great_expectations/core/configuration.py]: love it!

**Pull request merged**